### ISSUE TITLE
[Question] How to deal with source maps

### ISSUE THREAD

**Issue by MatteoGioioso** [microsoft/vscode-js-debug#901]
Hello, sorry if this is not an issue, but I have been trying to figure out for days how to deal with source maps and debugger.

My goal is to create a small debugger with limited functionality with the `inspector` API:

```
inspector.open(9229, 'localhost', false)

//... code to debug

inspector.close()
```

So basically I would send the command via websocket from another process.
It works all fine and I have achieved what I needed to do. However now I want to add source maps support for when the code is compiled. At the moment I am using compilation on the fly, to replicate the behaviour of the VSCode debugger:

```
node --require @babel/register --require @babel/polyfill --preserve-symlinks myTestScript.js
```
The test script includes ES6 modules that need to be compiled by babel.

### First implementation
So first I thought I could use the `source-map` library directly, however **the location given by the inspector in the scope chain or in the location object, never match the one from the source maps**. 
Moreover there is the problem to transform the current local scope into the original variable values, which I think is not possible

### Second implementation
I have tried to use `setScriptSource` on `scriptParsed` event however it does not seems to work:

```
            if (data.method === 'Debugger.scriptParsed') {
                    const sourceMapUrl = data.params.sourceMapURL

                    if (sourceMapUrl) {
                        const jsonSourceMaps = await SourceMapsExtractor.loadFromBase64(sourceMapUrl)
                        const originalSource = await SourceMapsExtractor.getOriginalFileSourceCode(originalFileUrl, jsonSourceMaps)
             
                        this._setScriptSource(data.params.scriptId, originalSource)
                    }
            }

```
`SourceMapExtractor` is a small wrapper around the mozilla library `source-map`.
So basically I would get the source url from the source map and get the original script code.

I would just like to know how it is implemented on your side and what is your approach to this problem, since online the material on this topic is very scarce.

I have also tried to go through the source code of this repo but I could not really get anything yet.

Thanks

**Comment by connor4312** [comment]
The inspector protocol never applies source maps for you. The approach with listening to `scriptParsed` and extracting the sourcemap is correct and is basically what js-debug does. You want to parse the map and associate with the script ID from the runtime. Then when you see locations in that script in the call stack, you use the `sourceMap.originalPositionFor` to map it back to the disk location.

Bear in mind that CDP is 0-based for line and column numbers, while the source-map library is 1-based for lines and 0-based for columns.

**Comment by MatteoGioioso** [comment]
@connor4312 thanks for the explanation. 
For "associate with the script ID from the runtime" you mean using `setScriptSource` ?

**Comment by connor4312** [comment]
no -- just keep something `Map<string /* script id */, SourceMap>` in your debugger

**Comment by MatteoGioioso** [comment]
I see, I will give it a try once more than.

Thanks for the help.
